File: authorization/src/main/kotlin/com/example/authorization/service/AuthorizationServiceImpl.java
package com.example.authorization.service;

import com.example.authorization.controller.*;
import com.example.authorization.entity.*;
import com.example.authorization.exception.*;
import com.example.authorization.repository.*;
import io.jsonwebtoken.*;
import io.jsonwebtoken.security.*;
import lombok.*;
import org.springframework.security.core.*;
import org.springframework.security.core.authority.*;
import org.springframework.security.crypto.password.*;
import org.springframework.stereotype.*;

import javax.crypto.*;
import java.nio.charset.*;
import java.util.*;
import java.util.stream.*;

import static com.example.authorization.constants.SecurityConstants.*;

@Service
@RequiredArgsConstructor
public class AuthorizationServiceImpl implements AuthorizationService {
    private final UserRepository userRepository;
    private final AuthoritiesRepository authoritiesRepository;
    private final PasswordEncoder passwordEncoder;

    @Override
    public User registerUser(UserRequest userRequest) {
        String username = userRequest.username();
        String password = userRequest.password();
        String role = userRequest.role();

        validateUserWithThatUsernameDoesNotExist(username);
        Authorities authority = getAuthority(role);
        User newUser = buildUser(username, password, authority);

        return userRepository.save(newUser);
    }

    @Override
    public String loginUser(UserRequest userRequest) {
        String username = userRequest.username();
        String password = userRequest.password();
        User user = getUserByUsername(username);
        validPasswords(password, user.getPassword());

        List<GrantedAuthority> authorities = Collections.singletonList(
                new SimpleGrantedAuthority(user.getAuthorities().getAuthority())
        );
        SecretKey key = createSecretKey();

        return buildJwt(username, authorities, key);
    }

    @Override
    public UserResponse getUserDetails(String jwt) {
        Claims claims = parseJwtClaims(jwt);
        String username = String.valueOf(claims.get("username"));
        String authorities = String.valueOf(claims.get("authorities"));
        return new UserResponse(username, authorities);
    }

    private User buildUser(String username, String password, Authorities authority) {
        return User.builder()
                .username(username)
                .password(passwordEncoder.encode(password))
                .authorities(authority)
                .build();
    }

    private String buildJwt(String username, List<GrantedAuthority> authorities, SecretKey key) {
        return Jwts.builder()
                .setIssuer("Social Media")
                .setSubject("JWT Token")
                .claim("username", username)
                .claim("authorities", populateAuthorities(authorities))
                .setIssuedAt(new Date())
                .setExpiration(new Date(JWT_EXPIRE_TIME))
                .signWith(key)
                .compact();
    }

    private SecretKey createSecretKey() {
        return Keys.hmacShaKeyFor(JWT_KEY.getBytes(StandardCharsets.UTF_8));
    }

    private void validPasswords(String passwordProvided, String passwordRegistered) {
        if (!isValidPasswords(passwordProvided, passwordRegistered)) {
            throw new WrongCredentialsException("Invalid password!");
        }
    }

    private void validateUserWithThatUsernameDoesNotExist(String username) {
        if (userRepository.existsByUsername(username)) {
            throw new UsernameAlreadyExistsException("User with this username already exists!");
        }
    }

    private boolean isValidPasswords(String passwordProvided, String passwordRegistered) {
        return passwordEncoder.matches(passwordProvided, passwordRegistered);
    }

    private User getUserByUsername(String username) {
        return userRepository.findUserByUsername(username)
                .orElseThrow(() -> new UserNotFoundException("No user registered with this username!"));
    }

    private Authorities getAuthority(String authority) {
        return authoritiesRepository.findAuthoritiesByAuthority(authority)
                .orElse(authoritiesRepository.findAuthoritiesByAuthority("ROLE_USER")
                        .orElseThrow(() -> new RuntimeException("No authorities found!")));
    }

    private Claims parseJwtClaims(String jwt) {
        SecretKey key = Keys.hmacShaKeyFor(JWT_KEY.getBytes(StandardCharsets.UTF_8));
        return Jwts.parserBuilder()
                .setSigningKey(key)
                .build()
                .parseClaimsJws(jwt)
                .getBody();
    }

    private String populateAuthorities(Collection<? extends GrantedAuthority> collection) {
        return collection.stream()
                .map(GrantedAuthority::getAuthority)
                .collect(Collectors.joining(","));
    }
}

File: authorization/src/main/kotlin/com/example/authorization/service/AuthorizationServiceImpl.kt
package com.example.authorization.service

import com.example.authorization.client.service.AuthorApiService
import com.example.authorization.controller.dto.*
import com.example.authorization.entity.*
import com.example.authorization.exception.*
import com.example.authorization.repository.*
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service
import java.time.Instant


@Service
class AuthorizationServiceImpl(
    private val userRepository: UserRepository,
    private val authoritiesRepository: AuthoritiesRepository,
    private val authorApiService: AuthorApiService,
    private val jwtService: JwtService
) : AuthorizationService {
    override fun registerUser(registerRequest: RegisterRequest) {
        val username = registerRequest.username
        val password = registerRequest.password
        val firstName = registerRequest.firstName
        val lastName = registerRequest.lastName
        val role = registerRequest.role

        validateUserWithThatUsernameDoesNotExist(username)
        val authority = getAuthority(role)
        val newUser = buildUser(username, password, authority)
        userRepository.save(newUser)
        val authorRequest = AuthorRequest(
            firstName,
            lastName,
            username
        )
        authorApiService.createAuthor(authorRequest)
    }

    @Transactional
    override fun deleteUser(jwt: String) {
        val user = getUserDetails(jwt)
        userRepository.deleteByUsername(user.username)
        authorApiService.deleteAuthorById(user.authorId)
    }

    override fun loginUser(loginRequest: LoginRequest): LoginResponse {
        val username = loginRequest.username
        val password = loginRequest.password
        val user = getUserByUsername(username)
        validPasswords(password, user.password)

        val authority = Authorities()
        authority.authority = user.authorities.authority
        val authorities = listOf(
            authority
        )

        val expirationDate = Instant.now().toEpochMilli() + 180000
        val token = jwtService.buildJwt(username, authorities, expirationDate)
        val author = authorApiService.getAuthorByUsername(username)

        return LoginResponse(
            username = username,
            firstName = author.firstName,
            lastName = author.lastName,
            token = token,
            tokenExpirationDate = expirationDate
        )
    }

    override fun getUserDetails(jwt: String): UserDetailsDTO {
        val claims = jwtService.parseJwtClaims(jwt)
        val username = claims["username"].toString()
        val authorities = claims["authorities"].toString()
        val author = authorApiService.getAuthorByUsername(username)

        return UserDetailsDTO(
            authorId = author.id,
            firstName = author.firstName,
            lastName = author.lastName,
            username = username,
            role = authorities
        )
    }

    private fun buildUser(username: String, password: String, authority: Authorities) =
        User(
            username,
            password,
            authority
        )

    private fun validPasswords(passwordProvided: String, passwordRegistered: String) {
        if (passwordProvided != passwordRegistered) {
            throw WrongCredentialsException("Invalid password!")
        }
    }

    private fun validateUserWithThatUsernameDoesNotExist(username: String) {
        if (userRepository.existsByUsername(username)) {
            throw UsernameAlreadyExistsException("User with this username already exists!")
        }
    }

    private fun getUserByUsername(username: String) =
        userRepository.findUserByUsername(username)
            ?: throw UserNotFoundException("No user registered with this username!")

    private fun getAuthority(authority: String) =
        authoritiesRepository.findAuthoritiesByAuthority(authority)
            ?: authoritiesRepository.findAuthoritiesByAuthority("ROLE_USER")
            ?: throw RuntimeException("No authorities found!")
}
